flight3handler.go

```diff
@@ -5,9 +5,24 @@ import (
 )
 
 func flight3Parse(ctx context.Context, c flightConn, state *State, cache *handshakeCache, cfg *handshakeConfig) (flightVal, *alert, error) {
-	var msgs map[handshakeType]handshakeMessage
-	var ok bool
-	var seq int
+	// Clients may receive multiple HelloVerifyRequest messages with different cookies.
+	// Clients SHOULD handle this by sending a new ClientHello with a cookie in response
+	// to the new HelloVerifyRequest. RFC 6347 Section 4.2.1
+	seq, msgs, ok := cache.fullPullMap(state.handshakeRecvSequence,
```

Comment:
What handshake message sequence number does the server use in the new `HelloVerifyRequest`?
The code currently assume that new `HelloVerifyRequest` has a sequence number continued from the previous `HelloVerifyRequest`.
In this case, `HelloVerifyRequest` will be duplicated in the `handshakeCache`.
I haven't fully checked yet, but I think duplicated `HelloVerifyRequest` in `handshakeCache` will cause `cache.pullAndMerge` result duplicated and make data signatures generated in flight4 and 5 wrong.

Reply:
> What handshake message sequence number does the server use in the new HelloVerifyRequest?

The first `HelloVerifyRequest` goes with `MessageSequence: 0` and the second one with `MessageSequence: 1`

Reply:
So, the handshake will be like
```
         Client                             Server
         ------                             ------
         ClientHello (seq=0)  ------>

                              <------ HelloVerifyRequest (seq=0)

         ClientHello (seq=1)  ------>
         (with cookie)

                              <------ HelloVerifyRequest (seq=1) (with new cookie)

         ClientHello (seq=2)  ------>
         (with new cookie)

                              <------        ServerHello (seq=2)
                              <------        Certificate (seq=3)
                              <------    ServerHelloDone (seq=4)

         [Rest of handshake]
```

> I haven't fully checked yet, but I think duplicated HelloVerifyRequest in handshakeCache will cause cache.pullAndMerge result duplicated and make data signatures generated in flight4 and 5 wrong.

This was wrong. `handshakeCache.pull()` only outputs the last message if same type of the messages are given multiple times. It satisfies https://tools.ietf.org/html/rfc6347#section-4.2.1
> If HelloVerifyRequest is used, the initial ClientHello and HelloVerifyRequest are not included in the calculation of the handshake_messages (for the CertificateVerify message) and verify_data (for the Finished message).

and `ClientHello (seq=0)`, `HelloVerifyRequest (seq=0)` and `ClientHello (seq=1)` will not be included in the MAC calculation. This behavior makes sense to me.

Reply:
Thanks for the analysis!